File: modules/core/src/main/java/com/panayotis/jubler/undo/UndoList.java
/*
 * (c) 2005-2023 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

package  com.panayotis.jubler.undo;

import com.panayotis.jubler.JubFrame;
import com.panayotis.jubler.subs.Subtitles;
import java.util.Stack;

public class UndoList extends Stack<UndoEntry> {

    private JubFrame jub;
    private UndoList redo;
    /* Mark which undo is what has been saved */
    private int unsaved_pos = 0;

    public UndoList(JubFrame j) {
        super();
        if (j == null)
            return;  // This is a REDO list
        jub = j;
        redo = new UndoList(null);
    }

    public void addUndo(UndoEntry undo) {
        if (undo == null)
            return;
        push(undo);
        jub.setDoText(undo.getName(), true);
        redo.removeAllElements();
        jub.setDoText(null, false);
        jub.resetUndoMark();
        jub.setUnsaved(true);
        jub.showInfo();

        /* We can't just call tableHasChanged here, because the table hasn't changed yet! */
        // jub.tableHasChanged();  
    }

    public void applyDoCommand(Subtitles subs, boolean isUndo, int[] selected_rows) {
        UndoEntry entry;
        UndoList source, dest;

        if (isUndo) {
            source = this;
            dest = redo;
        } else {
            source = redo;
            dest = this;
        }

        if (source.size() == 0)
            return;
        entry = source.pop();
        dest.push(entry);

        if (source.size() == 0)
            jub.setDoText(null, isUndo);
        else
            jub.setDoText(source.peek().getName(), isUndo);
        jub.setDoText(entry.getName(), !isUndo);

        jub.setUnsaved(size() != unsaved_pos);
        jub.setSubs(entry.flipSubtitles(subs));
        jub.setSelectedSub(selected_rows, true);
    }

    public void setSaveMark() {
        unsaved_pos = size();
        jub.setUnsaved(false);
    }

    public void invalidateSaveMark() {
        unsaved_pos = -1;
        jub.setUnsaved(true);
    }
}

File: src/core/undo/UndoList.cpp
/*
 * (c) 2005-2026 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

#include "core/undo/UndoList.h"

void UndoList::addUndo(std::unique_ptr<UndoEntry> entry) {
    if (!entry)
        return;
    undo_.append(std::shared_ptr<UndoEntry>(std::move(entry)));
    host_->setDoText(undo_.last()->getName(), true);
    redo_.clear();
    host_->setDoText(QString(), false);
    host_->resetUndoMark();
    host_->setUnsaved(true);
    host_->autoHideEncodingBar();  // first edit: drop the transient encoding bar and its byte buffer
    host_->showInfo();
}

void UndoList::applyDoCommand(bool isUndo, const QList<int> &selectedRows) {
    QList<std::shared_ptr<UndoEntry>> &source = isUndo ? undo_ : redo_;
    QList<std::shared_ptr<UndoEntry>> &dest = isUndo ? redo_ : undo_;
    if (source.isEmpty())
        return;
    std::shared_ptr<UndoEntry> entry = source.takeLast();
    dest.append(entry);
    host_->setDoText(source.isEmpty() ? QString() : source.last()->getName(), isUndo);
    host_->setDoText(entry->getName(), !isUndo);
    host_->setUnsaved(undo_.size() != unsavedPos_);
    // Swap the window document with the snapshot.
    std::unique_ptr<Subtitles> placeholder = std::make_unique<Subtitles>();
    std::unique_ptr<Subtitles> current = host_->swapSubtitles(std::move(placeholder));
    std::unique_ptr<Subtitles> restored = entry->flipSubtitles(std::move(current));
    host_->swapSubtitles(std::move(restored));
    host_->setSelectedRows(selectedRows, true);
}

void UndoList::setSaveMark() {
    unsavedPos_ = undo_.size();
    host_->setUnsaved(false);
}

void UndoList::invalidateSaveMark() {
    unsavedPos_ = -1;
    host_->setUnsaved(true);
}
